In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Station Data Analysis

In [5]:
station_df = pd.read_csv("/home/junchen/Downloads/Divvy_Stations_Trips_2014_Q3Q4/Divvy_Stations_2014-Q3Q4.csv")

In [6]:
station_df

id                         name   latitude  longitude  dpcapacity  \
0      5       State St & Harrison St  41.873958 -87.627739          19   
1     13   Wilton Ave & Diversey Pkwy  41.932500 -87.652681          19   
2     14          Morgan St & 18th St  41.858086 -87.651073          15   
3     15         Racine Ave & 19th St  41.856453 -87.656471          15   
4     16          Wood St & North Ave  41.910329 -87.672516          15   
..   ...                          ...        ...        ...         ...   
295  347       Ashland Ave & Grace St  41.950687 -87.668700          15   
296  348     California Ave & 21st St  41.854016 -87.695445          15   
297  349  Halsted St & Wrightwood Ave  41.929143 -87.649077          15   
298  350    Ashland Ave & Chicago Ave  41.895966 -87.667747          15   
299  351  Cottage Grove Ave & 51st St  41.803038 -87.606615          15   

          dateCreated  
0     6/10/2013 10:46  
1     6/22/2013 18:29  
2     6/22/2013 18:33  
3     6/22/2013 18:35  
4     6/22/2013 18:55  
..                ...  
295   10/4/2013 16:06  
296   10/9/2013 16:13  
297  10/10/2013 12:40  
298  10/10/2013 14:22  
299  10/10/2013 16:56  

[300 rows x 6 columns]

In [7]:
# General station information summarization
station_df.describe()

id    latitude   longitude  dpcapacity
count  300.000000  300.000000  300.000000  300.000000
mean   189.063333   41.896274  -87.648202   16.800000
std     99.484487    0.040992    0.023024    4.673987
min      5.000000   41.788746  -87.707857   11.000000
25%    108.750000   41.871814  -87.665824   15.000000
50%    196.500000   41.894611  -87.648305   15.000000
75%    276.250000   41.926397  -87.631846   19.000000
max    351.000000   41.978353  -87.580715   47.000000

## Q4 Trip Data Analysis

In [37]:
q4_trip_df = pd.read_csv("/home/junchen/Downloads/Divvy_Stations_Trips_2014_Q3Q4/Divvy_Trips_2014-Q4.csv")

In [38]:
q4_trip_df

trip_id         starttime          stoptime  bikeid  tripduration  \
0       4413167  12/31/2014 23:54  12/31/2014 23:57    1880           193   
1       4413166  12/31/2014 23:50  12/31/2014 23:52    1153           161   
2       4413165  12/31/2014 23:48     1/1/2015 0:38    1946          3002   
3       4413164  12/31/2014 23:46     1/1/2015 0:37    2563          3084   
4       4413163  12/31/2014 23:44     1/1/2015 0:10     389          1570   
...         ...               ...               ...     ...           ...   
437960  3810756    10/1/2014 0:01    10/1/2014 0:16    1122           853   
437961  3810754    10/1/2014 0:02    10/1/2014 0:10    1531           468   
437962  3810753    10/1/2014 0:02    10/1/2014 0:16    2707           836   
437963  3810752    10/1/2014 0:01    10/1/2014 0:05    2059           238   
437964  3810751    10/1/2014 0:00    10/1/2014 0:09    1316           538   

        from_station_id             from_station_name  to_station_id  \
0                   296        Broadway & Belmont Ave            334   
1                   264  Stetson Ave & South Water St             44   
2                   168        Michigan Ave & 14th St            168   
3                   168        Michigan Ave & 14th St            168   
4                    43  Michigan Ave & Washington St            163   
...                 ...                           ...            ...   
437960              253    Clifton Ave & Lawrence Ave            114   
437961              113     Bissell St & Armitage Ave            340   
437962              332    Halsted St & Diversey Pkwy            333   
437963               92       Carpenter St & Huron St             29   
437964               52        Michigan Ave & Lake St            255   

                    to_station_name    usertype gender  birthyear  
0       Lake Shore Dr & Belmont Ave  Subscriber   Male     1989.0  
1            State St & Randolph St  Subscriber   Male     1984.0  
2            Michigan Ave & 14th St    Customer    NaN        NaN  
3            Michigan Ave & 14th St  Subscriber   Male     1987.0  
4          Damen Ave & Clybourn Ave  Subscriber   Male     1983.0  
...                             ...         ...    ...        ...  
437960   Sheffield Ave & Addison St  Subscriber   Male     1980.0  
437961    Clark St & Wrightwood Ave  Subscriber   Male     1985.0  
437962   Ashland Ave & Blackhawk St  Subscriber   Male     1985.0  
437963     Noble St & Milwaukee Ave  Subscriber   Male     1986.0  
437964   Indiana Ave & Roosevelt Rd  Subscriber   Male     1980.0  

[437965 rows x 12 columns]

In [39]:
# General q3 trip data information summarization
q4_trip_df.describe()

trip_id         bikeid   tripduration  from_station_id  \
count  4.379650e+05  437965.000000  437965.000000    437965.000000   
mean   4.108582e+06    1543.139025     790.854043       160.252331   
std    1.727278e+05     873.505754    1599.697104        98.595454   
min    3.810751e+06       1.000000      60.000000         5.000000   
25%    3.960317e+06     773.000000     347.000000        74.000000   
50%    4.105442e+06    1563.000000     570.000000       150.000000   
75%    4.257789e+06    2306.000000     942.000000       241.000000   
max    4.413167e+06    3047.000000   86392.000000       351.000000   

       to_station_id      birthyear  
count  437965.000000  368411.000000  
mean      159.902844    1978.302057  
std        98.648069      10.623102  
min         5.000000    1898.000000  
25%        74.000000    1972.000000  
50%       146.000000    1981.000000  
75%       240.000000    1986.000000  
max       351.000000    1998.000000

#### Male type data analysis

In [40]:
# Check number of male in the dataframe and the relevant information
q4_trip_male = q4_trip_df[q4_trip_df["gender"] == "Male"]
q4_trip_male.describe()[["tripduration", "birthyear"]]

tripduration      birthyear
count  286381.000000  286376.000000
mean      617.036197    1977.869811
std      1121.363684      10.683574
min        60.000000    1898.000000
25%       311.000000    1971.000000
50%       492.000000    1981.000000
75%       767.000000    1986.000000
max     85250.000000    1998.000000

In [41]:
# check number of subscriber in the dataframe who are male
q4_trip_male_subscriber = q4_trip_male[q4_trip_male["usertype"] == "Subscriber"]
q4_trip_male_subscriber.describe()[["tripduration", "birthyear"]]

tripduration      birthyear
count  286360.000000  286355.000000
mean      616.958388    1977.869257
std      1121.347331      10.683768
min        60.000000    1898.000000
25%       311.000000    1971.000000
50%       492.000000    1981.000000
75%       767.000000    1986.000000
max     85250.000000    1998.000000

In [42]:
# Check number of customer in the dataframe who are male
q4_trip_male_customer = q4_trip_male[q4_trip_male["usertype"] == "Customer"]
q4_trip_male_customer.describe()[["tripduration", "birthyear"]]

tripduration    birthyear
count     21.000000    21.000000
mean    1678.047619  1985.428571
std      814.449107     0.507093
min      524.000000  1985.000000
25%     1261.000000  1985.000000
50%     1398.000000  1985.000000
75%     1742.000000  1986.000000
max     3843.000000  1986.000000

#### Female type data analysis

In [43]:
q4_trip_female = q4_trip_df[q4_trip_df["gender"] == "Female"]
q4_trip_female.describe()[["tripduration", "birthyear"]]

tripduration     birthyear
count  81995.000000  81995.000000
mean     740.921885   1979.826989
std     1496.035966     10.244544
min       60.000000   1930.000000
25%      365.000000   1975.000000
50%      583.000000   1983.000000
75%      926.000000   1987.000000
max    83969.000000   1998.000000

In [44]:
# check number of subscriber in the dataframe who are female
q4_trip_female_subscriber = q4_trip_female[q4_trip_female["usertype"] == "Subscriber"]
q4_trip_female_subscriber.describe()[["tripduration", "birthyear"]]

tripduration     birthyear
count  81991.000000  81991.000000
mean     740.896013   1979.826591
std     1496.063187     10.244635
min       60.000000   1930.000000
25%      365.000000   1975.000000
50%      583.000000   1983.000000
75%      926.000000   1987.000000
max    83969.000000   1998.000000

In [45]:
# Check number of customer in the dataframe who are male
q4_trip_female_customer = q4_trip_female[q4_trip_female["usertype"] == "Customer"]
q4_trip_female_customer.describe()[["tripduration", "birthyear"]]

tripduration  birthyear
count      4.000000        4.0
mean    1271.250000     1988.0
std      619.031165        0.0
min      518.000000     1988.0
25%      907.250000     1988.0
50%     1330.500000     1988.0
75%     1694.500000     1988.0
max     1906.000000     1988.0

### Q3 Dataframe with Nan Analysis

In [46]:
# As we can observe, there are some columns contain NaN information. Let's extract them.
is_nan = q4_trip_df.isnull()
row_has_nan = is_nan.any(axis=1)
rows_with_nan = q4_trip_df[row_has_nan]
rows_with_nan

trip_id         starttime          stoptime  bikeid  tripduration  \
2       4413165  12/31/2014 23:48     1/1/2015 0:38    1946          3002   
26      4413140  12/31/2014 22:00  12/31/2014 22:23    2203          1416   
27      4413139  12/31/2014 21:58  12/31/2014 22:23     475          1445   
32      4413133  12/31/2014 21:48  12/31/2014 21:53     583           321   
34      4413130  12/31/2014 21:45  12/31/2014 21:58    2203           826   
...         ...               ...               ...     ...           ...   
437925  3810799    10/1/2014 0:35    10/1/2014 0:52    1842          1059   
437930  3810792    10/1/2014 0:32    10/1/2014 0:49     563          1063   
437932  3810790    10/1/2014 0:31    10/1/2014 0:43    2423           712   
437947  3810773    10/1/2014 0:16    10/1/2014 0:25     807           511   
437948  3810771    10/1/2014 0:13    10/1/2014 0:25    1330           717   

        from_station_id                 from_station_name  to_station_id  \
2                   168            Michigan Ave & 14th St            168   
26                   52            Michigan Ave & Lake St             52   
27                   52            Michigan Ave & Lake St             52   
32                   90                   Millennium Park             52   
34                   90                   Millennium Park             52   
...                 ...                               ...            ...   
437925               50          Clark St & Congress Pkwy             19   
437930              220        Hampden Ct & Diversey Pkwy             84   
437932               55  Halsted St & James M Rochford St             55   
437947              318    Southport Ave & Irving Park Rd            253   
437948              318    Southport Ave & Irving Park Rd            316   

                         to_station_name  usertype gender  birthyear  
2                 Michigan Ave & 14th St  Customer    NaN        NaN  
26                Michigan Ave & Lake St  Customer    NaN        NaN  
27                Michigan Ave & Lake St  Customer    NaN        NaN  
32                Michigan Ave & Lake St  Customer    NaN        NaN  
34                Michigan Ave & Lake St  Customer    NaN        NaN  
...                                  ...       ...    ...        ...  
437925             Loomis St & Taylor St  Customer    NaN        NaN  
437930             Union Ave & Grand Ave  Customer    NaN        NaN  
437932  Halsted St & James M Rochford St  Customer    NaN        NaN  
437947        Clifton Ave & Lawrence Ave  Customer    NaN        NaN  
437948         Damen Ave & Sunnyside Ave  Customer    NaN        NaN  

[69594 rows x 12 columns]

In [16]:
rows_with_nan.describe()

trip_id        bikeid  tripduration  from_station_id  \
count  6.959400e+04  69594.000000  69594.000000     69594.000000   
mean   4.061244e+06   1542.321450   1564.939937       149.964120   
std    1.585330e+05    875.754255   2750.746651       101.060845   
min    3.810771e+06      1.000000     60.000000         5.000000   
25%    3.931981e+06    769.000000    744.250000        61.000000   
50%    4.056670e+06   1568.000000   1187.000000       122.000000   
75%    4.152248e+06   2308.000000   1675.000000       233.000000   
max    4.413165e+06   3047.000000  86392.000000       351.000000   

       to_station_id  birthyear  
count   69594.000000       40.0  
mean      147.806478     1947.0  
std       100.263316        0.0  
min         5.000000     1947.0  
25%        60.000000     1947.0  
50%       117.000000     1947.0  
75%       227.000000     1947.0  
max       351.000000     1947.0

In [18]:
# Check unique user type in the nan trip dataframe
unique_user_type_nan = set(list(rows_with_nan["usertype"]))
unique_user_type_nan

{'Customer', 'Subscriber'}

In [20]:
# Extract customer data information from nan dataframe
rows_with_nan_customer = rows_with_nan[rows_with_nan["usertype"] == "Customer"]
rows_with_nan_customer.describe()

trip_id        bikeid  tripduration  from_station_id  \
count  6.954900e+04  69549.000000  69549.000000     69549.000000   
mean   4.061230e+06   1542.257976   1565.026125       149.973486   
std    1.585098e+05    875.715085   2750.601862       101.079530   
min    3.810771e+06      1.000000     60.000000         5.000000   
25%    3.931986e+06    769.000000    745.000000        61.000000   
50%    4.056685e+06   1568.000000   1187.000000       122.000000   
75%    4.151835e+06   2308.000000   1676.000000       233.000000   
max    4.413165e+06   3047.000000  86392.000000       351.000000   

       to_station_id  birthyear  
count   69549.000000        0.0  
mean      147.809171        NaN  
std       100.278883        NaN  
min         5.000000        NaN  
25%        60.000000        NaN  
50%       117.000000        NaN  
75%       227.000000        NaN  
max       351.000000        NaN

In [21]:
# Extract subscriber data information from nan dataframe
rows_with_nan_subscriber = rows_with_nan[rows_with_nan["usertype"] == "Subscriber"]
rows_with_nan_subscriber.describe()

trip_id       bikeid  tripduration  from_station_id  \
count  4.500000e+01    45.000000     45.000000        45.000000   
mean   4.082271e+06  1640.422222   1431.733333       135.488889   
std    1.919790e+05   939.663394   2996.572522        65.276552   
min    3.815301e+06   147.000000    185.000000        25.000000   
25%    3.892432e+06   846.000000    421.000000       100.000000   
50%    4.041415e+06  1647.000000    722.000000       173.000000   
75%    4.270741e+06  2478.000000    905.000000       173.000000   
max    4.362184e+06  2940.000000  14836.000000       331.000000   

       to_station_id  birthyear  
count      45.000000       40.0  
mean      143.644444     1947.0  
std        72.998492        0.0  
min        24.000000     1947.0  
25%       100.000000     1947.0  
50%       173.000000     1947.0  
75%       212.000000     1947.0  
max       331.000000     1947.0

So, in total, there 69594 rows of trip data contain nan in the q3 trip datafame. The percentage takes around 15.8%. At the same time, most customers have nan information on their birth year and gender.

### Q3 DataFrame Non-NaN Data Analysis

In [47]:
# Similarly, extract rows contains no nan
is_not_nan = q4_trip_df.notnull()
row_no_nan = is_not_nan.all(axis=1)
rows_no_nan = q4_trip_df[row_no_nan]
rows_no_nan

trip_id         starttime          stoptime  bikeid  tripduration  \
0       4413167  12/31/2014 23:54  12/31/2014 23:57    1880           193   
1       4413166  12/31/2014 23:50  12/31/2014 23:52    1153           161   
3       4413164  12/31/2014 23:46     1/1/2015 0:37    2563          3084   
4       4413163  12/31/2014 23:44     1/1/2015 0:10     389          1570   
5       4413162  12/31/2014 23:30  12/31/2014 23:44     588           831   
...         ...               ...               ...     ...           ...   
437960  3810756    10/1/2014 0:01    10/1/2014 0:16    1122           853   
437961  3810754    10/1/2014 0:02    10/1/2014 0:10    1531           468   
437962  3810753    10/1/2014 0:02    10/1/2014 0:16    2707           836   
437963  3810752    10/1/2014 0:01    10/1/2014 0:05    2059           238   
437964  3810751    10/1/2014 0:00    10/1/2014 0:09    1316           538   

        from_station_id             from_station_name  to_station_id  \
0                   296        Broadway & Belmont Ave            334   
1                   264  Stetson Ave & South Water St             44   
3                   168        Michigan Ave & 14th St            168   
4                    43  Michigan Ave & Washington St            163   
5                   210     Ashland Ave & Division St            153   
...                 ...                           ...            ...   
437960              253    Clifton Ave & Lawrence Ave            114   
437961              113     Bissell St & Armitage Ave            340   
437962              332    Halsted St & Diversey Pkwy            333   
437963               92       Carpenter St & Huron St             29   
437964               52        Michigan Ave & Lake St            255   

                       to_station_name    usertype gender  birthyear  
0          Lake Shore Dr & Belmont Ave  Subscriber   Male     1989.0  
1               State St & Randolph St  Subscriber   Male     1984.0  
3               Michigan Ave & 14th St  Subscriber   Male     1987.0  
4             Damen Ave & Clybourn Ave  Subscriber   Male     1983.0  
5       Southport Ave & Wellington Ave  Subscriber   Male     1988.0  
...                                ...         ...    ...        ...  
437960      Sheffield Ave & Addison St  Subscriber   Male     1980.0  
437961       Clark St & Wrightwood Ave  Subscriber   Male     1985.0  
437962      Ashland Ave & Blackhawk St  Subscriber   Male     1985.0  
437963        Noble St & Milwaukee Ave  Subscriber   Male     1986.0  
437964      Indiana Ave & Roosevelt Rd  Subscriber   Male     1980.0  

[368371 rows x 12 columns]

In [17]:
rows_no_nan.describe()

trip_id         bikeid   tripduration  from_station_id  \
count  3.683710e+05  368371.000000  368371.000000    368371.000000   
mean   4.117526e+06    1543.293484     644.610898       162.196017   
std    1.738389e+05     873.081413    1215.897586        98.001642   
min    3.810751e+06       1.000000      60.000000         5.000000   
25%    3.968832e+06     774.000000     322.000000        74.000000   
50%    4.123657e+06    1563.000000     510.000000       154.000000   
75%    4.267886e+06    2306.000000     802.000000       241.000000   
max    4.413167e+06    3047.000000   85250.000000       351.000000   

       to_station_id      birthyear  
count  368371.000000  368371.000000  
mean      162.188134    1978.305456  
std        98.172821      10.618670  
min         5.000000    1898.000000  
25%        75.000000    1972.000000  
50%       153.000000    1981.000000  
75%       241.000000    1986.000000  
max       351.000000    1998.000000

In [22]:
# Check unique user type in the non-nan trip dataframe
unique_user_type_nonan = set(list(rows_no_nan["usertype"]))
unique_user_type_nonan

{'Customer', 'Subscriber'}

In [23]:
# Extract customer data information from non-nan dataframe
rows_no_nan_customer = rows_no_nan[rows_no_nan["usertype"] == "Customer"]
rows_no_nan_customer.describe()

trip_id       bikeid  tripduration  from_station_id  \
count  2.500000e+01    25.000000     25.000000        25.000000   
mean   4.082163e+06  1537.160000   1612.960000       173.280000   
std    3.311685e+04   835.586355    789.835768       112.323758   
min    4.049136e+06     4.000000    518.000000        42.000000   
25%    4.057005e+06  1040.000000   1258.000000        53.000000   
50%    4.070257e+06  1572.000000   1398.000000       173.000000   
75%    4.111481e+06  2145.000000   1742.000000       273.000000   
max    4.139744e+06  2985.000000   3843.000000       345.000000   

       to_station_id    birthyear  
count      25.000000    25.000000  
mean      138.920000  1985.840000  
std        96.063312     1.067708  
min        43.000000  1985.000000  
25%        45.000000  1985.000000  
50%       140.000000  1986.000000  
75%       177.000000  1986.000000  
max       340.000000  1988.000000

In [24]:
# Extract subscriber data information from non-nan dataframe
rows_no_nan_subscriber = rows_no_nan[rows_no_nan["usertype"] == "Subscriber"]
rows_no_nan_subscriber.describe()

trip_id         bikeid   tripduration  from_station_id  \
count  3.683460e+05  368346.000000  368346.000000    368346.000000   
mean   4.117528e+06    1543.293900     644.545175       162.195265   
std    1.738443e+05     873.084988    1215.895960        98.000731   
min    3.810751e+06       1.000000      60.000000         5.000000   
25%    3.968818e+06     774.000000     322.000000        74.000000   
50%    4.123670e+06    1563.000000     510.000000       154.000000   
75%    4.267897e+06    2306.000000     802.000000       241.000000   
max    4.413167e+06    3047.000000   85250.000000       351.000000   

       to_station_id      birthyear  
count  368346.000000  368346.000000  
mean      162.189713    1978.304944  
std        98.172903      10.618845  
min         5.000000    1898.000000  
25%        75.000000    1972.000000  
50%       153.000000    1981.000000  
75%       241.000000    1986.000000  
max       351.000000    1998.000000

## Station-Trip Data Analysis

In [66]:
import datetime

In [58]:
q4_trip_df_from_station = q4_trip_df["from_station_id"]
from_id_count = q4_trip_df_from_station.value_counts()[:20]
from_id_count

91     9059
174    6987
192    6795
75     5401
35     5269
195    5109
283    4798
287    4792
49     4267
81     4250
76     4185
77     4152
48     3956
90     3936
176    3921
110    3909
100    3879
52     3874
66     3780
37     3725
Name: from_station_id, dtype: int64

In [59]:
q4_trip_df_to_station = q4_trip_df["to_station_id"]
to_id_count = q4_trip_df_to_station.value_counts()[:20]
to_id_count

91     9957
174    8268
192    7055
35     6641
283    5154
75     5098
287    4990
90     4821
43     4214
37     4168
52     4141
110    4092
211    4044
176    4031
81     4025
100    3965
77     3954
49     3946
76     3884
48     3783
Name: to_station_id, dtype: int64

In [65]:
# Both from trip and to trip ids contain 91, 174, 192 as the top three stations. 
#So, I will select the first three of them
q4_trip_selected = q4_trip_df[(q4_trip_df["from_station_id"] == 91) |
                              (q4_trip_df["from_station_id"] == 174) |
                              (q4_trip_df["from_station_id"] == 192)]

q4_trip_selected_91 = q4_trip_selected[q4_trip_selected["from_station_id"] == 91]
q4_trip_selected_174 = q4_trip_selected[q4_trip_selected["from_station_id"] == 174]
q4_trip_selected_192 = q4_trip_selected[q4_trip_selected["from_station_id"] == 192]

In [74]:
q4_trip_selected_to_station = q4_trip_df[(q4_trip_df["to_station_id"] == 91) |
                              (q4_trip_df["to_station_id"] == 174) |
                              (q4_trip_df["to_station_id"] == 192)]

q4_trip_selected_91_to_station = q4_trip_selected_to_station[q4_trip_selected_to_station["to_station_id"] == 91]
q4_trip_selected_174_to_station = q4_trip_selected_to_station[q4_trip_selected_to_station["to_station_id"] == 174]
q4_trip_selected_192_to_station = q4_trip_selected_to_station[q4_trip_selected_to_station["to_station_id"] == 192]

In [72]:
station_91 = station_df[station_df["id"] == 91]
station_174 = station_df[station_df["id"] == 174]
station_192 = station_df[station_df["id"] == 192]

### Average time duration that bike is borrowed from a station

In [68]:
# Station 91 trip duration: from station
q4_trip_selected_91.describe()["tripduration"]

count     9059.000000
mean       624.908047
std        764.361098
min         61.000000
25%        386.000000
50%        534.000000
75%        734.000000
max      30850.000000
Name: tripduration, dtype: float64

In [69]:
# Station 174 trip duration: from station
q4_trip_selected_174.describe()["tripduration"]

count     6987.000000
mean       642.121225
std        878.948607
min         63.000000
25%        358.000000
50%        499.000000
75%        767.500000
max      37508.000000
Name: tripduration, dtype: float64

In [70]:
# Station 192 trip duration: from station
q4_trip_selected_192.describe()["tripduration"]

count     6795.000000
mean       647.101251
std        999.551623
min         60.000000
25%        374.000000
50%        541.000000
75%        761.000000
max      50142.000000
Name: tripduration, dtype: float64

### Average time bike is sent back to station

In [75]:
# Station 91 trip duration: to station
q4_trip_selected_91_to_station.describe()["tripduration"]

count     9957.000000
mean       593.486492
std        754.168331
min         65.000000
25%        377.000000
50%        513.000000
75%        706.000000
max      46015.000000
Name: tripduration, dtype: float64

In [76]:
# Station 174 trip duration: to station
q4_trip_selected_174_to_station.describe()["tripduration"]

count     8268.000000
mean       569.401427
std       1282.797318
min         61.000000
25%        323.000000
50%        445.000000
75%        632.000000
max      67639.000000
Name: tripduration, dtype: float64

In [87]:
# Station 192 trip duration: to station
q4_trip_selected_192_to_station.describe()["tripduration"]

count     7055.000000
mean       633.718923
std       1258.949153
min         60.000000
25%        381.000000
50%        563.000000
75%        739.500000
max      63182.000000
Name: tripduration, dtype: float64

### Final: Detection of dock shortage

In [101]:
def dock_shortage(df, station_id):
    out = []
    count = 0
    for ids in station_id:
        from_station_info = df[df["from_station_id"] == ids].describe()["tripduration"][:2]
        to_station_info = df[df["to_station_id"] == ids].describe()["tripduration"][:2]
        if ((to_station_info[0] < from_station_info[0]) & (from_station_info[1] < to_station_info[1])):
            out.append(ids)
#         count += 1    
#         print(count)
    return out

shortage_list = dock_shortage(q4_trip_df, station_df["id"])

In [102]:
shortage_list

[13,
 16,
 21,
 24,
 29,
 30,
 45,
 46,
 47,
 49,
 50,
 54,
 66,
 68,
 75,
 76,
 77,
 92,
 97,
 98,
 111,
 114,
 116,
 143,
 146,
 153,
 154,
 164,
 165,
 169,
 175,
 178,
 179,
 185,
 190,
 191,
 194,
 195,
 197,
 200,
 202,
 204,
 207,
 213,
 214,
 215,
 217,
 225,
 232,
 233,
 238,
 241,
 242,
 246,
 247,
 249,
 252,
 253,
 255,
 256,
 259,
 264,
 271,
 273,
 274,
 275,
 276,
 280,
 286,
 301,
 305,
 309,
 312,
 314,
 316,
 333,
 335,
 340,
 341,
 342,
 343,
 349,
 350]

In [109]:
print("There are in total {} number of dock station are in shortage.".format(len(shortage_list)))

There are in total 83 number of dock station are in shortage.


### Final: Enough Dock Station Detection

In [104]:
def dock_enough(df, station_id):
    out = []
    count = 0
    for ids in station_id:
        from_station_info = df[df["from_station_id"] == ids].describe()["tripduration"][:2]
        to_station_info = df[df["to_station_id"] == ids].describe()["tripduration"][:2]
        if ((to_station_info[0] > from_station_info[0]) & (from_station_info[1] > to_station_info[1])):
            out.append(ids)
#         count += 1    
#         print(count)
    return out

enough_list = dock_enough(q4_trip_df, station_df["id"])

In [105]:
enough_list

[5,
 15,
 22,
 26,
 27,
 31,
 34,
 35,
 37,
 43,
 52,
 57,
 67,
 85,
 86,
 90,
 91,
 100,
 108,
 110,
 113,
 121,
 122,
 123,
 128,
 129,
 130,
 140,
 141,
 144,
 156,
 162,
 166,
 168,
 170,
 171,
 173,
 174,
 176,
 181,
 184,
 186,
 192,
 193,
 198,
 201,
 211,
 212,
 216,
 218,
 219,
 220,
 227,
 230,
 234,
 239,
 240,
 244,
 245,
 250,
 261,
 283,
 285,
 287,
 289,
 293,
 298,
 300,
 302,
 306,
 307,
 313,
 317,
 318,
 319,
 323,
 324,
 325,
 327,
 329,
 331,
 334,
 336,
 346]

In [108]:
print("There are total {} number of enough dock stations".format(len(enough_list)))

There are total 84 number of enough dock stations
